cmd/juju/action/call.go

```diff
@@ -0,0 +1,310 @@
+}
+
+func (c *callCommand) Run(ctx *cmd.Context) error {
+	if err := c.ensureAPI(); err != nil {
+		return errors.Trace(err)
+	}
+	defer c.api.Close()
+
+	actionParams := map[string]interface{}{}
+	if c.paramsYAML.Path != "" {
+		b, err := c.paramsYAML.Read(ctx)
+		if err != nil {
+			return err
+		}
+
+		err = yaml.Unmarshal(b, &actionParams)
+		if err != nil {
+			return err
+		}
+
+		conformantParams, err := common.ConformYAML(actionParams)
+		if err != nil {
+			return err
+		}
+
+		betterParams, ok := conformantParams.(map[string]interface{})
+		if !ok {
+			return errors.New("params must contain a YAML map with string keys")
+		}
+
+		actionParams = betterParams
+	}
+
+	// If we had explicit args {..., [key, key, key, key, value], ...}
+	// then iterate and set params ..., key.key.key.key=value, ...
+	for _, argSlice := range c.args {
+		valueIndex := len(argSlice) - 1
+		keys := argSlice[:valueIndex]
+		value := argSlice[valueIndex]
+		cleansedValue := interface{}(value)
+		if !c.parseStrings {
+			err := yaml.Unmarshal([]byte(value), &cleansedValue)
+			if err != nil {
+				return err
+			}
+		}
+		// Insert the value in the map.
+		addValueToMap(keys, cleansedValue, actionParams)
+	}
+
+	conformantParams, err := common.ConformYAML(actionParams)
+	if err != nil {
+		return err
+	}
+
+	typedConformantParams, ok := conformantParams.(map[string]interface{})
+	if !ok {
+		return errors.Errorf("params must be a map, got %T", typedConformantParams)
+	}
+
+	actions := make([]params.Action, len(c.unitReceivers))
+	for i, unitReceiver := range c.unitReceivers {
+		if strings.HasSuffix(unitReceiver, "leader") {
+			if c.api.BestAPIVersion() < 3 {
+				app := strings.Split(unitReceiver, "/")[0]
+				return errors.Errorf("unable to determine leader for application %q"+
+					"\nleader determination is unsupported by this API"+
+					"\neither upgrade your controller, or explicitly specify a unit", app)
+			}
+			actions[i].Receiver = unitReceiver
+		} else {
+			actions[i].Receiver = names.NewUnitTag(unitReceiver).String()
+		}
+		actions[i].Name = c.actionName
+		actions[i].Parameters = actionParams
+	}
+	results, err := c.api.Enqueue(params.Actions{Actions: actions})
+	if err != nil {
+		return err
+	}
+
+	if len(results.Results) != len(c.unitReceivers) {
+		return errors.New("illegal number of results returned")
+	}
+
+	for _, result := range results.Results {
+		if result.Error != nil {
+			return result.Error
+		}
+		if result.Action == nil {
+			return errors.Errorf("action failed to enqueue on %q", result.Action.Receiver)
+		}
+		tag, err := names.ParseActionTag(result.Action.Tag)
+		if err != nil {
+			return err
+		}
+
+		// Legacy Juju 1.25 output format for a single unit, no wait.
+		if !c.wait.forever && c.wait.d.Nanoseconds() <= 0 && len(results.Results) == 1 {
+			out := map[string]string{"Action queued with id": tag.Id()}
+			return c.out.Write(ctx, out)
+		}
+	}
+
+	out := make(map[string]interface{}, len(results.Results))
```

Comment:
This name is so close to c.out... Can it be named something else, say 'results'?